Redesign search filtering experience to be consistent with mobile

myreli opened the pull request:
Hi! I hope you are doing well today. :grin: 

This PR provides a redesign for the filtering options. The goal is to provide a more consistent experience cross platform, bringing the search options as bubbles instead of switches on a modal. 

https://user-images.githubusercontent.com/17554234/156862312-812ca54c-e8ea-40c5-9d12-474fb043354e.mp4

## Summary
Redesigned the `SearchOptions` component to create a similar experience to the one we have on mobile when filtering:
- Removed the search options modal and previous options
- Changed the search trigger to `OnFocus` instead of modal
- Created the new headless `Bubble` component
- Removed the previous `onBlur` effect from Search to ensure a more consistent flow with the new experience

## Approach
I inspected the experience provided on mobile and did my best to replicate on web, given the difference in usability. 

Probably the biggest difference is that I've renamed `"Include protected [Option]"` to simply `"[Option]"` to improve readability and avoid horizontal scroll (when possible). Do you think that could mislead users?

## Thoughts
- :warning: I don't have Front End experience so I would appreciate an extra look on CSS and React specifics. 
- :thought_balloon: I realized this components aren't tested, except for `NoteView` that contains heavy logic. Is there any other place I should consider for e2e or perhaps am I missing something?
- :test_tube: I created a small collection of notes to test it, if you wish to try it out importing [simulated-data.txt](https://github.com/standardnotes/web/files/8189567/simulated-data.txt)

---

Thank you for your review!

amanharwara commented:
Hi @myreli, thanks for the contribution! :rocket: You're not missing anything regarding the tests, btw. For the React and CSS, there are a few things that need to be adjusted, but looks good overall :slightly_smiling_face::+1:

amanharwara commented on app/assets/stylesheets/_notes.scss:
```diff
@@ -71,6 +72,39 @@
+
+        transition-property: color, background-color, border-color, transform;
+        transition-timing-function: cubic-bezier(.4,0,.2,1);
+        transition-duration: 300ms;
```
Looking at the video, I think this transition is a bit slow. Something like `100ms` might look better, what do you think?

myreli replied:
Agree! Thank you for the suggestion. :grin:

myreli commented on app/assets/stylesheets/_notes.scss:
```diff
@@ -71,6 +72,39 @@
+
+      .bubble {
+        cursor: pointer;
+        padding: 2px 5px;
```
Thank you! I have [used the utility](https://github.com/standardnotes/web/pull/908/commits/105592826382898fc0dcaa0ed7d781a48dd6a1eb#diff-2ec334de391c3dc55ad45bfd28b7de9e421725252c82b13bd130911e9b43d9aeR8) classes to increase and use `rem`.

myreli commented on app/assets/stylesheets/_notes.scss:
```diff
@@ -71,6 +72,39 @@
+      font-size: var(--sn-stylekit-font-size-p);
+      white-space: nowrap;
+
+      .bubble {
```
Great! I've changed to utility classes on https://github.com/standardnotes/web/pull/908/commits/fdecd9c410c74534e8fbc2d0dc959c685224ba40. 

I knew TailwindCSS, bus wasn't familiar with it. Thank you for sharing, really helped! Is SN planning on moving on to it? I see the [optimization engine](https://tailwindcss.com/docs/optimizing-for-production) that removes unused CSS looks very promising.

amanharwara replied:
> Is SN planning on moving on to it? I see the [optimization engine](https://tailwindcss.com/docs/optimizing-for-production) that removes unused CSS looks very promising.

Maybe in the future. But for now the plan is to eventually move utility classes from this repo to [StyleKit](https://github.com/standardnotes/StyleKit) where part of our existing utility classes already are, so that editors can use them as well.

myreli commented:
Hi @amanharwara! Thank you for the review. 

I have just updated the PR with https://github.com/standardnotes/web/pull/908/commits/fdecd9c410c74534e8fbc2d0dc959c685224ba40, that improves the code and result with yours and @moughxyz's suggestions. Could you check again when you have the time?

Here's a **summary** of the changes: 

- Animated the search options appearance and improved the existing transition
- Increased the padding of each bubble
- Improved the spacing of the search options
- Fixed a regression on the input search borders
- Refactored the CSS to use utility classes instead of component specifics ¹

https://user-images.githubusercontent.com/17554234/156907679-9cd63924-d863-469a-a487-5ffdd62f6cb3.mp4

¹ I tried to keep the new utility classes next to the similar existing. However I was unsure when choosing `sn.scss` instead of `ui.scss`.

amanharwara commented:
> However I was unsure when choosing `sn.scss` instead of `ui.scss`.

`sn.scss` is where the utility classes should go.

myreli commented:
Hey @amanharwara! Thank you again. 

Since the video is failing, here's how it looks so far:
![image](https://user-images.githubusercontent.com/17554234/156928622-57839f9e-528e-4f80-9ec2-b9da3123d129.png)

amanharwara commented:
Looks good! One small issue with this is that since this is using justify-between, if the user increases the size of the notes column, those would be spread too far apart. I would suggest using regular margins or the `gap` property instead so that they don't get too far apart or too close to each other.

myreli commented:
Got it @amanharwara. I increased gap and justified on center. 
![image](https://user-images.githubusercontent.com/17554234/156929539-98aa9196-cf0b-4e75-89be-0398db966818.png)

amanharwara reviewed (approved):
Looks good!

amanharwara commented:
> Probably the biggest difference is that I've renamed `"Include protected [Option]"` to simply `"[Option]"` to improve readability and avoid horizontal scroll (when possible). Do you think that could mislead users?

I was giving this a go locally and I think this might actually end up being a bit confusing, as I found out myself. Having just "Protected" conveys that you are filtering based on whether the note is protected or not (like the Archived/Trashed options), instead of conveying that the user will be toggling whether to search contents of protected files or not. It might be better to have the whole "Include Protected Contents" label, which might make sizing tricky, but will be better in terms of clarity.
